File: Objective C/TestProject/ViewController.m
//
//  ViewController.m
//  TestProject
//
//  Created by Admin on 10/16/16.
//  Copyright © 2016 Nirav Bhatt. All rights reserved.
//

#import "ViewController.h"
#import "CustomObject.h"
#import "TableSearchViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (IBAction)loadNumbers:(id)sender
{
    NSArray * firstSectionArray = @[@(1.23), @(2.45), @(3.0)];
    NSArray * secondSectionArray = @[@(2), @(5), @(6)];
    
    NSArray * resultsArray = @[@{@"First Section" : firstSectionArray}, @{@"Second Section" : secondSectionArray}];
    
    TableSearchViewController * tableSearchViewController = [[TableSearchViewController alloc] initWithCellColorStyle:CELL_COLOR_STYLE_ALTERNATE_DOUBLE andSectionColorStyle:SECTION_COLOR_STYLE_UNIFORM andAllowSelectionCheckMark:NO andAllowSelectAllCheckBox:YES andAllowSearch:NO andAccessoryAction:ACCESSORY_ACTION_CHECK andFooterText:@"This is extra footer checkbox" andResultsArray:resultsArray];
    
    tableSearchViewController.textLableFormats = @[@"My numbers: %.4f"];
    tableSearchViewController.preventSubTitle = YES;
    
    tableSearchViewController.selectionDoneBlock = ^(NSArray* _Nullable  selectedKVCObjects, BOOL bExtraFlag)
    {
        NSLog(@"%@", selectedKVCObjects);
        _valueLabel.text = [NSString stringWithFormat:@"Selected: %@", [selectedKVCObjects componentsJoinedByString:@","]];
    };
    
    UINavigationController * navCtrl = [[UINavigationController alloc] initWithRootViewController:tableSearchViewController];
    [self presentViewController:navCtrl animated:YES completion:nil];
}

- (IBAction)loadObjects:(id)sender
{
    CustomObject * firstObj = [[CustomObject alloc] initWithString:@"First" andNumber:@(1)];
    CustomObject * secondObj = [[CustomObject alloc] initWithString:@"Second" andNumber:@(2.555)];
    CustomObject * thirdObj = [[CustomObject alloc] initWithString:@"Third" andNumber:@(3.999)];
    
    CustomObject * forthObj = [[CustomObject alloc] initWithString:@"Four" andNumber:@(4.111)];
    CustomObject * fifthObj = [[CustomObject alloc] initWithString:@"Five" andNumber:@(5.345)];
    
    NSArray * firstSectionArray = @[firstObj, secondObj, thirdObj];
    NSArray * secondSectionArray = @[forthObj, fifthObj, firstObj];
    
    NSArray * resultsArray = @[@{@"First Section" : firstSectionArray}, @{@"Second Section" : secondSectionArray}];
    
    TableSearchViewController * tableSearchViewController = [[TableSearchViewController alloc] initWithCellColorStyle:CELL_COLOR_STYLE_ALTERNATE_DOUBLE andSectionColorStyle:SECTION_COLOR_STYLE_UNIFORM andAllowSelectionCheckMark:NO andAllowSelectAllCheckBox:YES andAllowSearch:YES andAccessoryAction:ACCESSORY_ACTION_CHECK andFooterText:@"" andResultsArray:resultsArray];
    
    //Must define below properties for seachability
    tableSearchViewController.searchKeys = @[@"stringProperty"];
    tableSearchViewController.textLabelKeys = @[@"stringProperty"];
    tableSearchViewController.subTitleKeys = @[@"numberProperty"];
    tableSearchViewController.textLableFormats = @[@"String Property is: %@"];
    tableSearchViewController.subTitleFormats = @[@"Number Property is: %.4f"];
    
    tableSearchViewController.selectionDoneBlock = ^(NSArray* _Nullable  selectedKVCObjects, BOOL bExtraFlag)
    {
        NSLog(@"%@", selectedKVCObjects);
        _valueLabel.text = [NSString stringWithFormat:@"Selected: %@", [selectedKVCObjects componentsJoinedByString:@","]];
    };
    
    UINavigationController * navCtrl = [[UINavigationController alloc] initWithRootViewController:tableSearchViewController];
    [self presentViewController:navCtrl animated:YES completion:nil];
}

- (IBAction)loadStrings:(id)sender
{
    NSArray * firstSectionArray = @[@"one", @"two", @"three"];
    NSArray * secondSectionArray = @[@"one", @"three", @"five",  @"six",];
    
    NSArray * resultsArray = @[@{@"First Section" : firstSectionArray}, @{@"Second Section" : secondSectionArray}];
    
    TableSearchViewController * tableSearchViewController = [[TableSearchViewController alloc] initWithCellColorStyle:CELL_COLOR_STYLE_ALTERNATE_DOUBLE andSectionColorStyle:SECTION_COLOR_STYLE_UNIFORM andAllowSelectionCheckMark:NO andAllowSelectAllCheckBox:YES andAllowSearch:YES andAccessoryAction:ACCESSORY_ACTION_CHECK andFooterText:@"" andResultsArray:resultsArray];
    
    tableSearchViewController.checkBoxText = @"This is the footer checkbox.";
    tableSearchViewController.selectionDoneBlock = ^(NSArray* _Nullable  selectedKVCObjects, BOOL bExtraFlag)
    {
        NSLog(@"%@", selectedKVCObjects);
        _valueLabel.text = [NSString stringWithFormat:@"Selected: %@", [selectedKVCObjects componentsJoinedByString:@","]];
    };
    
    UINavigationController * navCtrl = [[UINavigationController alloc] initWithRootViewController:tableSearchViewController];
    [self presentViewController:navCtrl animated:YES completion:nil];
}

- (IBAction)loadStringsWithDelete:(id)sender
{
    NSArray * firstSectionArray = @[@"one", @"two", @"two", @"three"];
    NSArray * secondSectionArray = @[@"one", @"three", @"five",  @"six",];
    
    NSArray * resultsArray = @[@{@"First Section" : firstSectionArray}, @{@"Second Section" : secondSectionArray}];
    
    TableSearchViewController * tableSearchViewController = [[TableSearchViewController alloc] initWithCellColorStyle:CELL_COLOR_STYLE_ALTERNATE_DOUBLE andSectionColorStyle:SECTION_COLOR_STYLE_UNIFORM andAllowSelectionCheckMark:NO andAllowSelectAllCheckBox:YES andAllowSearch:YES andAccessoryAction:ACCESSORY_ACTION_DELETE andFooterText:@"" andResultsArray:resultsArray];
    
    tableSearchViewController.selectionDoneBlock = ^(NSArray* _Nullable  deletedKVCObjects, BOOL bExtraFlag)
    {
        NSLog(@"%@", deletedKVCObjects);
        _valueLabel.text = [NSString stringWithFormat:@"%@", [deletedKVCObjects componentsJoinedByString:@","]];
    };
    
    tableSearchViewController.accessoryActionDoneBlock = ^(NSArray* _Nullable  accessoryActionKVCObjects)
    {
        NSLog(@"%@", accessoryActionKVCObjects);
        _valueLabel.text = [NSString stringWithFormat:@"Deleted: %@", [accessoryActionKVCObjects componentsJoinedByString:@","]];
    };
    
    UINavigationController * navCtrl = [[UINavigationController alloc] initWithRootViewController:tableSearchViewController];
    [self presentViewController:navCtrl animated:YES completion:nil];
}


@end

File: Swift/TableSearchViewController/ViewController.swift
//
//  ViewController.swift
//  TableSearchViewController
//
//  Created by Admin on 5/1/18.
//  Copyright © 2018 iPhoneGameZone. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var resultLabel: UILabel!
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }

    @IBAction func loadStrings(_ sender: Any)
    {
        let resultsArray = [["First": ["one","two","three"]], ["Second": ["two","four","six","seven"]]]

        let tableVC = TableSearchViewController.init(cellColorStyle: CellColorStyle.CELL_COLOR_STYLE_UNIFORM, sectionColorStyle: SectionColorStyle.SECTION_COLOR_STYLE_UNIFORM, allowSelectionCheckMark: false, allowSearch: true, accessoryAction: ACCESSORY_ACTION.ACCESSORY_ACTION_CHECK, footerText: "This is extra footer", resultsArray: resultsArray)
        
        //Specify what happens when "Select" is pressed and TableSearchViewController is dismissed
        tableVC.selectionDoneBlock = { selectedObjects, bExtraFlag in
            
            let str = "Selected: " + selectedObjects.description
            self.resultLabel.text? = str
            print("Selected:\(str)")
        }
        
        let tableNavVC  = UINavigationController.init(rootViewController: tableVC)
        
        self.navigationController?.present(tableNavVC, animated: true, completion: {
        })
    }
    
    @IBAction func loadNumbers(_ sender: Any)
    {
        let resultsArray = [["First": [1,2,3]], ["Second": [2,5,6,7]]]
        
        //Displays TableSearchViewController with checkmarks
        let tableVC = TableSearchViewController.init(cellColorStyle: CellColorStyle.CELL_COLOR_STYLE_UNIFORM, sectionColorStyle: SectionColorStyle.SECTION_COLOR_STYLE_UNIFORM, allowSelectionCheckMark: true, allowSearch: true, accessoryAction: ACCESSORY_ACTION.ACCESSORY_ACTION_CHECK, footerText: "This is extra footer", resultsArray: resultsArray)
        
        
        //Specify what happens when "Select" is pressed and TableSearchViewController is dismissed
        tableVC.selectionDoneBlock = { selectedObjects, bExtraFlag in
            
            let str = "Selected: " + selectedObjects.description
            self.resultLabel.text? = str
            print("Selected:\(str)")
        }
        
        let tableNavVC  = UINavigationController.init(rootViewController: tableVC)
        
        self.navigationController?.present(tableNavVC, animated: true, completion: {
        })
    }
    
    @IBAction func loadDictionaries (_ sender: Any)
    {
        let dict1 = ["text": "Totaldrunk", "num" : 1] as [String : AnyObject]
        let dict2 = ["text": "Lovelystrings", "num" : 2] as [String : AnyObject]
        let dict3 = ["text": "Flyingsaucer", "num" : 3] as [String : AnyObject]
        let dict4 = ["text": "lovelyapple", "num" : 4] as [String : AnyObject]
        let dict5 = ["text": "Flyingberries", "num" : 5] as [String : AnyObject]
        let dict6 = ["text": "kiterunner", "num" : 6] as [String : AnyObject]

        
        let resultsArray = [["First": [dict1, dict2, dict3]], ["Second": [dict3, dict4, dict5, dict6]]]
        
        let tableVC = TableSearchViewController.init(cellColorStyle: CellColorStyle.CELL_COLOR_STYLE_UNIFORM, sectionColorStyle: SectionColorStyle.SECTION_COLOR_STYLE_UNIFORM, allowSelectionCheckMark: false, allowSearch: true, accessoryAction: ACCESSORY_ACTION.ACCESSORY_ACTION_CHECK, footerText: "This is extra footer", resultsArray: resultsArray)
        
        
        //Specify what forms the basis for search
        tableVC.searchKeys = ["text"]
        
        //Specify what appears in titles
        tableVC.textLabelKeys = ["text"]
        
        //Specify what appears in subtitles
        tableVC.subTitleKeys = ["num"]
        
        //Specify how title is displayed - each format specifier preceded by % gets replaced by textLabelKey value
        tableVC.textLabelFormats = ["text is here: %@"]
        
        //Specify how subtitle is displayed - each format specifier preceded by % gets replaced by subTitleKey value
        tableVC.subTitleFormats = ["num is here: %@"]
        
        //Specify what happens when "Select" is pressed and TableSearchViewController is dismissed
        tableVC.selectionDoneBlock = { selectedObjects, bExtraFlag in
            let str = "Selected: " + selectedObjects.description
            self.resultLabel.text? = str
            print("Selected:\(str)")
        }

        let tableNavVC  = UINavigationController.init(rootViewController: tableVC)

        self.navigationController?.present(tableNavVC, animated: true, completion: {
        })
    }
    
    @IBAction func loadStringsWithDelete(_ sender: Any)
    {
        let resultsArray = [["First": ["one","two","three"]], ["Second": ["two","four","six","seven"]]]
        
        let tableVC = TableSearchViewController.init(cellColorStyle: CellColorStyle.CELL_COLOR_STYLE_UNIFORM, sectionColorStyle: SectionColorStyle.SECTION_COLOR_STYLE_UNIFORM, allowSelectionCheckMark: false, allowSearch: true, accessoryAction: ACCESSORY_ACTION.ACCESSORY_ACTION_DELETE, footerText: "", resultsArray: resultsArray)
        
        //Specify what happens when "Done" is pressed and TableSearchViewController is dismissed
        tableVC.selectionDoneBlock = { deletedObjects, bExtraFlag in
            
            let str = "All Deleted: " + (deletedObjects as Array<Any>).description
            self.resultLabel.text? = str
            print("Deleted:\(str)")
        }
        
        tableVC.deletionDoneBlock = { kvcObjectDeleted in
            let str = "Deleted Now: " + (kvcObjectDeleted as! String)
            self.resultLabel.text? = str
            print("Deleted:\(str)")
        }

        let tableNavVC  = UINavigationController.init(rootViewController: tableVC)
        
        self.navigationController?.present(tableNavVC, animated: true, completion: {
        })
    }
   
    @IBAction func loadNumbersWithDelete(_ sender: Any)
    {
        let resultsArray = [["First": [1,2,3]], ["Second": [2,5,6,7]]]
        
        //Displays TableSearchViewController with checkmarks
        let tableVC = TableSearchViewController.init(cellColorStyle: CellColorStyle.CELL_COLOR_STYLE_UNIFORM, sectionColorStyle: SectionColorStyle.SECTION_COLOR_STYLE_UNIFORM, allowSelectionCheckMark: true, allowSearch: true, accessoryAction: ACCESSORY_ACTION.ACCESSORY_ACTION_DELETE, footerText: "This is extra footer", resultsArray: resultsArray)
        
        
        //Specify what happens when "Done" is pressed and TableSearchViewController is dismissed
        tableVC.selectionDoneBlock = { deletedObjects, bExtraFlag in
            let str = "All Deleted: " + (deletedObjects as Array<Any>).description
            self.resultLabel.text? = str
            print("Deleted:\(str)")
        }
        
        tableVC.deletionDoneBlock = { kvcObjectDeleted in
            let str = "Deleted Now: " + String(kvcObjectDeleted as! Double)
            self.resultLabel.text? = str
            print("Deleted:\(str)")
        }
        
        let tableNavVC  = UINavigationController.init(rootViewController: tableVC)
        
        self.navigationController?.present(tableNavVC, animated: true, completion: {
        })
    }
    
    @IBAction func loadDictionariesWithDelete(_ sender: Any)
    {
        let dict1 = ["text": "Totaldrunk", "num" : 1] as [String : AnyObject]
        let dict2 = ["text": "Lovelystrings", "num" : 2] as [String : AnyObject]
        let dict3 = ["text": "Flyingsaucer", "num" : 3] as [String : AnyObject]
        let dict4 = ["text": "lovelyapple", "num" : 4] as [String : AnyObject]
        let dict5 = ["text": "Flyingberries", "num" : 5] as [String : AnyObject]
        let dict6 = ["text": "kiterunner", "num" : 6] as [String : AnyObject]
        
        let resultsArray = [["First": [dict1, dict2, dict3]], ["Second": [dict3, dict4, dict5, dict6]]]
        
        let tableVC = TableSearchViewController.init(cellColorStyle: CellColorStyle.CELL_COLOR_STYLE_UNIFORM, sectionColorStyle: SectionColorStyle.SECTION_COLOR_STYLE_UNIFORM, allowSelectionCheckMark: false, allowSearch: true, accessoryAction: ACCESSORY_ACTION.ACCESSORY_ACTION_DELETE, footerText: "This is extra footer", resultsArray: resultsArray)
        
        //Specify what forms the basis for search
        tableVC.searchKeys = ["text"]
        
        //Specify what appears in titles
        tableVC.textLabelKeys = ["text"]
        
        //Specify what appears in subtitles
        tableVC.subTitleKeys = ["num"]
        
        //Specify how title is displayed - each format specifier preceded by % gets replaced by textLabelKey value
        tableVC.textLabelFormats = ["text is here: %@"]
        
        //Specify how subtitle is displayed - each format specifier preceded by % gets replaced by subTitleKey value
        tableVC.subTitleFormats = ["num is here: %@"]
        
        //Specify what happens when "Done" is pressed and TableSearchViewController is dismissed
        tableVC.selectionDoneBlock = { deletedObjects, bExtraFlag in
            
            var str = ""
            for dict in deletedObjects
            {
                let dictionary = dict as! Dictionary<String, Any>
              
                let dictString = dictionary.keys.description + ":" + dictionary.values.description
                
                str.append(dictString)
            }
            
            let labelStr = "All Deleted: " + str
            self.resultLabel.text? = labelStr
            print("Deleted:\(labelStr)")
        }
        
        tableVC.deletionDoneBlock = { kvcObjDeleted  in
            let dictionary = kvcObjDeleted as! Dictionary<String, Any>
            let str = "Deleted Now: " + dictionary.keys.description + ":" + dictionary.values.description
            self.resultLabel.text? = str
            print("Deleted:\(str)")
        }
        
        let tableNavVC  = UINavigationController.init(rootViewController: tableVC)
        
        self.navigationController?.present(tableNavVC, animated: true, completion: {
        })
    }
    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()

    }
}
